geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/XZ2SFC.scala

```diff
@@ -0,0 +1,392 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.curve
+
+import org.locationtech.geomesa.curve.XZ2SFC.{QueryWindow, XElement}
+import org.locationtech.sfcurve.IndexRange
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+  * Extended Z-order curve implementation used for efficiently storing polygons.
+  *
+  * Based on 'XZ-Ordering: A Space-Filling Curve for Objects with Spatial Extension'
+  * by Christian Böhm, Gerald Klump  and Hans-Peter Kriegel
+  *
+  * @param g resolution level of the curve - i.e. how many times the space will be recursively quartered
+  */
+class XZ2SFC(g: Short) {
+
+  // TODO see what the max value of g can be where we can use Ints instead of Longs and possibly refactor to use Ints
+
+  /**
+    * Index a polygon by it's bounding box
+    *
+    * @param bounds (xmin, ymin, xmax, ymax)
+    * @return
+    */
+  def index(bounds: (Double, Double, Double, Double)): Long = index(bounds._1, bounds._2, bounds._3, bounds._4)
+
+  /**
+    * Index a polygon by it's bounding box
+    *
+    * @param xmin min x value in [-180,180]
+    * @param ymin min y value in [-90,90]
+    * @param xmax max x value in [-180,180], must be >= xmin
+    * @param ymax max y value in [-90,90], must be >= ymin
+    * @return
+    */
+  def index(xmin: Double, ymin: Double, xmax: Double, ymax: Double): Long = {
+    // normalize inputs to [0,1]
+    val (nxmin, nymin, nxmax, nymax) = XZ2SFC.normalize(xmin, ymin, xmax, ymax)
+
+    // calculate the length of the sequence code (section 4.1 of XZ-Ordering paper)
+
+    val w = nxmax - nxmin
+    val h = nymax - nymin
+
+    // el-one is a bit confusing to read, but corresponds with the paper's definitions
+    val l1 = math.min(g, math.floor(math.log(math.max(w, h)) / math.log(0.5)).toInt)
+
+    // predicate for checking how many axis the polygon intersects
+    def predicate(value: Double, wh: Double): Boolean = math.floor((value / l1) + 2) * l1 <= value + wh
```

Comment:
This predicate doesn't make sense to me. It's only true if l1 is 0 (which happens when `math.max(w, h)` is > .5). As such, there are several cases where the length is calculated at a finer resolution than it should be.